bee-network/src/transport.rs

```diff
@@ -0,0 +1,31 @@
+// Copyright 2020 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::config::DEFAULT_CONNECTION_TIMEOUT_SECS;
+
+use libp2p::{
+    core::{
+        muxing::StreamMuxerBox,
+        transport::{upgrade, Boxed},
+    },
+    dns, identity, mplex, noise, tcp, PeerId, Transport,
+};
+
+use std::io;
+
+pub fn build_transport(local_keys: &identity::Keypair) -> io::Result<Boxed<(PeerId, StreamMuxerBox)>> {
+    let noise_keys = noise::Keypair::<noise::X25519Spec>::new()
+        .into_authentic(local_keys)
+        .expect("error creating noise keys");
+
+    let tcp = tcp::TokioTcpConfig::new().nodelay(true);
+    let transport = dns::DnsConfig::new(tcp)?;
+
+    Ok(transport
+        .upgrade(upgrade::Version::V1)
+        .authenticate(noise::NoiseConfig::xx(noise_keys).into_authenticated())
+        .multiplex(mplex::MplexConfig::new())
+        // .multiplex(SelectUpgrade::new(yamux::Config::default(), mplex::MplexConfig::new()))
```

Comment:
Unused?

Reply:
unfortunately I had to do this, because `yamux` created an issue with `libp2p` where the network layer couldn't establish a transport layer with a peer using `yamux` as its mulitplexer. Only with `Mplex` the code worked as expected. We should, however, come back to this issue at a later point in time, and investigate. That's why I left this line out-commented in there. Once this issue is fixed, instead of thinking long about how to provide both muxers again, I can just uncomment this line and remove the previous.

Reply:
The reason why I think this is probably something that should be fixed by one of our dependencies is that `libp2p` understands itself as an abstraction layer over the concrete multiplexer being used, or in other words: my user code should depend on the multiplexer I use.